### PR TITLE
Preserve TypeLambdas in wildApprox

### LINKED ISSUE #6238: wildApprox clobbers HKTypeLambdas representing kinds
`wildApprox`'s job is to "approximate occurrences of parameter types and uninstantiated typevars by wildcard types". Unfortunately it's a little over-enthusiastic and also eliminates the type parameters bound by `HKTypeLambdas` used to represent kinds.

Implicit resolution is one of the main users of `wildApprox` and this bug showed up as a consequence of the [fix](https://github.com/lampepfl/dotty/pull/6163) for #6058. In the following,

```scala
class Foo[T]

class Bar[F[_]]
object Bar {
  implicit def barF[F[_]](implicit fooF: Foo[Bar[F]]): Bar[F] = null
}

class A[T]
object A {
  implicit def fooA[F[_[_]]](implicit barB: F[B]): Foo[F[A]] = null
}

class B[T]
object B {
  implicit def fooB[F[_[_]]]: Foo[F[B]] = null
}

implicitly[Bar[A]]
```
the divergence checker reports divergence (via `barF`) when expanding `Bar[B]` in the process of expanding `Bar[A]`. It does this because even though `Bar[A]` and `Bar[B]` are not `=:=` (which should block divergence) currently `wildApprox(Bar[A])` and `wildApprox(Bar[B])` _are_ `=:=`.

This happens because the types `A` and `B` in `Bar[A]` and `Bar[B]` are wrapped in `HKTypeLambdas` at this point in typechecking, ie. these types are effectively represented as,

```scala
Bar[[T] => A[T]]
Bar[[T] => B[T]]
```
When `wildApprox` is applied to them they're transformed into,
```scala
Bar[[T] => A[_]]
Bar[_ <: [_$1] => Any]
```
respectively, which `=:=` then report as equal (this is also a bit fishy, but the `wildApprox` transformation is clearly wrong).

This can be fixed (PR incoming) by modifying `wildApprox` to leave alone `TypeParamRefs` corresponding to `HKTypeLambdas` which represent kinds.

### PR DESCRIPTION
**Pull request by milessabin**
`wildApprox` should not replace `TypeParamRefs` which are bound by `TypeLambdas` which are contained in the type being approximated.

Fixes #6238.

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
This looks like its on the right track. Let's see what the performance tests say.

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala [hunk 1]
```diff
@@ -630,33 +631,54 @@ object ProtoTypes {
...
-          wildApprox(tp.argType, theMap, seen),
-          wildApprox(tp.resultType, theMap, seen))
+          wildApprox(tp.argType, theMap, seen, internal),
+          wildApprox(tp.resultType, theMap, seen, internal))
     case  _: ThisType | _: BoundType => // default case, inlined for speed
       tp
+    case tl: TypeLambda =>
+      val internal1 = internal + tl
+      val paramInfos = tl.paramInfos
+      val paramInfos1 = tl.paramInfos.mapConserve {
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala]: Can drop the `tl.` here.

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala [hunk 2]
```diff
@@ -630,33 +631,54 @@ object ProtoTypes {
...
+      val res1 = wildApprox(res, theMap, seen, internal1)
+      if ((res eq res1) && (paramInfos eq paramInfos1)) tl
+      else {
+        def substBounds(tl1: TypeLambda)(tb: TypeBounds): TypeBounds =
+          tb.derivedTypeBounds(tb.lo.subst(tl, tl1), tb.hi.subst(tl, tl1))
+        tl match {
+          case _: HKTypeLambda => HKTypeLambda(tl.paramNames)(tl1 => paramInfos.map(substBounds(tl1)), tl1 => res1.subst(tl, tl1))
+          case _: PolyType     => PolyType(tl.paramNames)(tl1 => paramInfos.map(substBounds(tl1)), tl1 => res1.subst(tl, tl1))
+        }
+      }
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala]: I *think* all this machinery is already provided by `derivedLambdaType`. Can you try this instead as the case for `TypeLambda`?
```scala
  val internal1 = internal + tl
  val paramInfos1 = tl.paramInfos.mapConserve {
    case tb @ TypeBounds(lo, hi) =>
          tb.derivedTypeBounds(
            wildApprox(lo, theMap, seen, internal1),
            wildApprox(hi, theMap, seen, internal1)
          )
      }
  val res1 = wildApprox(tl.resultType, theMap, seen, internal1)
  tl.derivedLambdaType(paramInfos = paramInfos1, resType = res1)
```

**Comment by odersky** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala (hunk 2 above)
Comment (milessabin) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala]: Damn. That's what I expected, and I looked for it, but somehow managed to miss `newLikeThis` and then ended up rewriting it inline :-(

Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala]: It can even be simplified further, it seems. See 85c641f.

**Comment by milessabin** [comment]
@odersky I'm fine with your simplifications wrt `HKTypeLambda`, but I don't understand why it's safe to handle `PolyTypes` in the way you're suggesting.

**Comment by odersky** [comment]
> @odersky I'm fine with your simplifications wrt HKTypeLambda, but I don't understand why it's safe to handle PolyTypes in the way you're suggesting.

We always reduce PolyTypes to their result type, inferring type parameters. Since the type parameter references of PolyTypes are eliminated anyway by wildApprox, there's no need to keep the parameters around. That's the thinking behind it, anyway.

**Comment by milessabin** [comment]
> Since the type parameter references of PolyTypes are eliminated anyway by wildApprox, there's no need to keep the parameters around

Right, but these type parameters are being bound _internally_ to the type being approximated, so do they really need to be wildcarded out? My understand was that the point of the approximation was only to eliminated references to type parameters bound _externally_ to the type being approximated.

**Comment by odersky** [comment]
> Right, but these type parameters are being bound internally to the type being approximated, so do they really need to be wildcarded out? My understand was that the point of the approximation was only to eliminate references to type parameters bound externally to the type being approximated.

I think it could go either way, but I did not find an instance where we would actually call `wildApprox` on a `PolyType`. So I agree, it's better to simplify and treat PolyTypes just like other TypeLambdas.

**Comment by milessabin** [comment]
This LGTM ... happy with this being merged if you are :-)

**Pull request merged**